File: go/teller.go
package main

import (
	"crypto/tls"
	"crypto/x509"
	"encoding/base64"
	"io"
	"log"
	"net/http"
	"net/http/httputil"
	"os"
	"path/filepath"
	"strings"
)

var (
	appID       = os.Getenv("APP_ID")
	environment = getEnv("ENV", "sandbox")
	certPath    = os.Getenv("CERT")
	keyPath     = os.Getenv("CERT_KEY")
	port        = getEnv("PORT", "8001")
)

func getEnv(key, fallback string) string {
	if v := os.Getenv(key); v != "" {
		return v
	}
	return fallback
}

func main() {
	if appID == "" {
		log.Fatal("APP_ID must be set (e.g. APP_ID=app_xxx)")
	}
	if (environment == "development" || environment == "production") &&
		(certPath == "" || keyPath == "") {
		log.Fatalf("CERT and CERT_KEY must be set when ENV=%s", environment)
	}

	mux := http.NewServeMux()
	staticDir := filepath.Join("..", "static")

	// Root: serve index.html with replacements
	mux.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		data, err := os.ReadFile(filepath.Join(staticDir, "index.html"))
		if err != nil {
			http.Error(w, "index.html not found", 500)
			return
		}
		html := string(data)
		html = strings.Replace(html, "{{ app_id }}", appID, 1)
		html = strings.Replace(html, "{{ environment }}", environment, 1)
		w.Header().Set("Content-Type", "text/html")
		io.WriteString(w, html)
	})

	// Static assets
	mux.Handle("/static/", http.StripPrefix("/static/", http.FileServer(http.Dir(staticDir))))

	// Proxy for /api/*
	mux.Handle("/api/", newTellerProxy())

	log.Printf("Listening on http://localhost:%s (ENV=%s, APP_ID=%s)", port, environment, appID)
	log.Fatal(http.ListenAndServe("0.0.0.0:"+port, mux))
}

func newTellerProxy() http.Handler {
	// Setup TLS if needed
	var tlsConfig *tls.Config
	if certPath != "" && keyPath != "" {
		cert, err := tls.LoadX509KeyPair(certPath, keyPath)
		if err != nil {
			log.Fatalf("failed to load cert/key: %v", err)
		}
		tlsConfig = &tls.Config{
			Certificates: []tls.Certificate{cert},
			RootCAs:      x509.NewCertPool(),
		}
	}

	proxy := &httputil.ReverseProxy{
		Director: func(req *http.Request) {
			// Rewrite URL: strip /api
			req.URL.Scheme = "https"
			req.URL.Host = "api.teller.io"
			req.URL.Path = strings.TrimPrefix(req.URL.Path, "/api")

			// Fix Authorization header: Basic token:
			if token := req.Header.Get("Authorization"); token != "" {
				encoded := base64.StdEncoding.EncodeToString([]byte(token + ":"))
				req.Header.Set("Authorization", "Basic "+encoded)
			}
		},
		ModifyResponse: func(resp *http.Response) error {
			// Strip hop-by-hop headers only
			for _, h := range []string{"Connection", "Transfer-Encoding"} {
				resp.Header.Del(h)
			}
			// Keep Content-Encoding so the browser can decompress gzip
			return nil
		},
	}
	if tlsConfig != nil {
		proxy.Transport = &http.Transport{TLSClientConfig: tlsConfig}
	}
	return proxy
}

File: elixir/teller.exs
Mix.install([
  {:plug_cowboy, "~> 2.6"},
  {:httpoison, "~> 2.2"}
])

defmodule Teller.Router do
  use Plug.Router

  plug :match
  plug :dispatch

  @app_id System.get_env("APP_ID") || raise "APP_ID must be set"
  @env    System.get_env("ENV") || "sandbox"
  @cert   System.get_env("CERT")
  @key    System.get_env("CERT_KEY")

  if @env in ["development", "production"] and (!@cert or !@key) do
    raise "CERT and CERT_KEY must be set when ENV=#{@env}"
  end

  # -------- Root (inject {{ app_id }} / {{ environment }}) --------
  get "/" do
    static_dir = Path.expand("../static", __DIR__)
    html =
      File.read!(Path.join(static_dir, "index.html"))
      |> String.replace("{{ app_id }}", @app_id, global: false)
      |> String.replace("{{ environment }}", @env,    global: false)

    conn
    |> put_resp_content_type("text/html")
    |> send_resp(200, html)
  end

  # -------- Static --------
  get "/static/*path" do
    static_dir = Path.expand("../static", __DIR__)
    file = Path.join([static_dir | path])

    if File.exists?(file), do: send_file(conn, 200, file), else: send_resp(conn, 404, "not found")
  end

  # -------- Proxy /api/* -> https://api.teller.io/* --------
  match "/api/*path" do
    teller_url = "https://api.teller.io/" <> Enum.join(path, "/")

    # Rewrite Authorization: <token> -> Basic base64("token:")
    headers =
      case get_req_header(conn, "authorization") do
        [token] -> [{"authorization", "Basic " <> Base.encode64(token <> ":")} | drop_auth(conn.req_headers)]
        _       -> drop_auth(conn.req_headers)
      end

    # Read body once
    {:ok, body, _} = Plug.Conn.read_body(conn)

    # TLS client certs (dev/prod)
    hackney_opts =
      if @cert && @key, do: [ssl_options: [certfile: @cert, keyfile: @key]], else: []

    # Forward
    method = conn.method |> String.to_atom()
    resp   = HTTPoison.request!(method, teller_url, body, headers, hackney: hackney_opts)

    resp_headers =
      resp.headers
      |> Enum.reject(fn {k, _} -> k in ["connection", "transfer-encoding"] end)

    conn
    |> put_resp_headers(resp_headers)
    |> send_resp(resp.status_code, resp.body)
  end

  match _ do
    send_resp(conn, 404, "not found")
  end

  defp put_resp_headers(conn, headers),
    do: Enum.reduce(headers, conn, fn {k, v}, acc -> Plug.Conn.put_resp_header(acc, k, v) end)

  defp drop_auth(headers),
    do: Enum.reject(headers, fn {k, _} -> k == "authorization" end)
end

require Logger

port = String.to_integer(System.get_env("PORT") || "8001")
Logger.info("Listening on http://localhost:#{port} (ENV=#{System.get_env("ENV") || "sandbox"}, APP_ID=#{System.get_env("APP_ID")})")

{:ok, _} = Plug.Cowboy.http(Teller.Router, [], port: port)
Process.sleep(:infinity)